### ISSUE TITLE
Bug: [no-unnecessary-condition] reports on indirect eval

### ISSUE THREAD

**Issue by Josh-Cena** [typescript-eslint/typescript-eslint#8990]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#ts=5.2.2&showAST=es&fileType=.tsx&code=KYNwhgNg-AdAFAIgB4IJQG4g&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
eval?.("x");
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No error. This is the shortest form to trigger indirect eval. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval#direct_and_indirect_eval

### Actual Result

Errors

**Comment by bradzacher** [comment]
The rule is powered by types - the type of `eval` is non-nullish, so it's an unnecessary condition.
Additionally use of eval is generally considered a code smell.

Additionally indirect eval doesn't work here if you're downlevelling your TS. The optional chain is transpiled to `eval === null || eval === void 0 ? void 0 : eval("x + y");` - which is a direct eval.

I don't think there's value to encoding these semantics into the rule.
This is a one in a million edge-case.

**Comment by Josh-Cena** [comment]
@bradzacher I know that it's technically "optional chain on non-nullish value", but this is the idiom to trigger indirect eval. I'm not downleveling and that downleveled output is not spec-compliant. Babel does it right: `eval === null || eval === void 0 ? void 0 : (0, eval)("x");`

Fare enough that it's rare, but it's clearly not "unnecessary".

**Comment by bradzacher** [comment]
> but it's clearly not "unnecessary"

```ts
let x = false;
new Promise(() => {
    x = true
});
if (x) {
  // unnecessary condition reported because `x` is typed as `false`
}
```

This is a common theme and is a known limitation of the rule - we intentionally are powered by types and respect what the types tell us.

---

I'd put it out there that one in a million devs have written code that uses eval; one in a hundred million know of indirect vs direct eval; and one in a billion have ever intentionally written code that uses indirect eval and even fewer still have done it via optional chaining.

It's just not worth the code to even think about this usecase.

---

- [TS gets it wrong](https://www.typescriptlang.org/play/?#code/KYNwhgNg-AdAFAIgB4IJQG4CwAoIA)
- [SWC gets it right accidentally](https://play.swc.rs/?version=1.3.100&code=H4sIAAAAAAAAA0stS8yx19NQqlDStObl0tdXyCxRL1ZITE7OTEnNK0nMyalUKMpMzyjh5UpEKAMATuvH7jUAAAA%3D&config=H4sIAAAAAAAAA1WPSw7DIAwF9zkF8rrbdtE79BAWdSIifrKJVBTl7iUE0maH3xsz8jooBbNoeKq1PMsQkYX4nEsi2Sf8lARIOxTNJia49XaWvRrRCtVoOxpIyBOluiX3hoMNQajjLXPGmzH%2FC3VwkUnkCu4o%2BsnSVTc0JbjwXmrZDkk50qF%2FwA%2FqsvNjMPLqm4kXGrYvhlQioBQBAAA%3D)
- [esbuild gets it right](https://esbuild.github.io/try/#dAAwLjIwLjIAeyB0YXJnZXQ6ICJFUzUiIH0AZXZhbD8uKCJ4Iik7Cg)

**Comment by kirkwaiblinger** [comment]
I think you should have to write about 18 eslint-disable coments to use direct or indirect eval. 😆

**Comment by Josh-Cena** [comment]
> * TS gets it wrong

TS does a lot of downleveling wrong. It's an explicitly non-compliant compiler.

SWC does not get it right "by accident"; it was first fixed in https://github.com/swc-project/swc/issues/2735 but apparently they later decided all optional chaining should use the same syntax, which is fine too.

> This is a common theme and is a known limitation of the rule - we intentionally are powered by types and respect what the types tell us.

Again I'm just saying the types are absolutely correct and reflect the runtime value—not arguing with that. I'm just saying here the optional chaining has a semantic other than "short-circuit for nullish".

Agree that eval is used rarely, but I wish when people use it they can use it correctly by default (indirect is the correct default) and not have to be penalized by unrelated lint rules as if they did something suspicious.
